### PR TITLE
Add healthchecking and optimize system-agent

### PR DESCRIPTION
**Pull request by Oats87**
https://github.com/rancher/system-agent/issues/8

### CODE REVIEW

File: pkg/applyinator/applyinator.go [hunk 1]
```diff
@@ -28,6 +29,38 @@ type Applyinator struct {
 	imageUtil       *image.Utility
 }
 
+// CalculatedPlan is passed into Applyinator and is a Plan with checksum calculated
+type CalculatedPlan struct {
+	Plan     Plan
```
Comment (briandowns) [inline pkg/applyinator/applyinator.go:34]: Will this value be mutated by the applyinator?

File: pkg/applyinator/applyinator.go [hunk 2]
```diff
@@ -43,22 +76,42 @@ func NewApplyinator(workDir string, preserveWorkDir bool, appliedPlanDir string,
 	}
 }
 
-func (a *Applyinator) Apply(ctx context.Context, anp types.AgentNodePlan) ([]byte, error) {
-	logrus.Infof("Applying plan with checksum %s", anp.Checksum)
+func CalculatePlan(rawPlanByteData []byte) (CalculatedPlan, error) {
```
Comment (briandowns) [inline pkg/applyinator/applyinator.go:79]: Can you remove the Hungarian notation? It provides an unnecessary redundancy.

File: pkg/k8splan/watcher.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+				probePeriod = parsedPeriod
+			}
+		}
+		if secret == nil {
+			logrus.Debugf("[K8s] Secret was nil")
+			core.Secret().EnqueueAfter(w.connInfo.Namespace, w.connInfo.SecretName, probePeriod)
+			return secret, nil
+		}
+		secret = secret.DeepCopy()
+		logrus.Debugf("[K8s] Processing secret %s in namespace %s at generation %d with resource version %s", secret.Name, secret.Namespace, secret.Generation, secret.ResourceVersion)
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:89]: Would it be clearer if we changed `[K8s]` to `[system-agent]`?

File: pkg/k8splan/watcher.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+				}
+			}
+
+			var wg sync.WaitGroup
+			var mu sync.Mutex
+
+			for probeName, probe := range cp.Plan.Probes {
+				wg.Add(1)
+
+				go func() {
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:146]:
```suggestion
				go func(wg *sync.WaitGroup) {
```

File: pkg/k8splan/watcher.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+						probeStatus = prober.ProbeStatus{}
+					}
+					if err := prober.DoProbe(probe, &probeStatus, needsApplied); err != nil {
+						logrus.Errorf("error running probe %s", probeName)
+					}
+					mu.Lock()
+					logrus.Debugf("[K8s] (%s) writing probe status to map", probeName)
+					probeStatuses[probeName] = probeStatus
+					mu.Unlock()
+				}()
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:164]:
```suggestion
				}(&wg)
```

File: pkg/k8splan/watcher.go (hunk 3 above)
Comment (Oats87) [reply pkg/k8splan/watcher.go:89]: The reason for `[K8s]` is to help distinguish between `[K8s]` and `[local]` (source of plan data)

File: pkg/applyinator/applyinator.go (hunk 2 above)
Comment (Oats87) [reply pkg/applyinator/applyinator.go]: Hungarian notation removed

File: pkg/applyinator/applyinator.go (hunk 1 above)
Comment (Oats87) [reply pkg/applyinator/applyinator.go:34]: This value generally is not mutated by the applyinator

File: pkg/k8splan/watcher.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+const (
+	appliedChecksumKey   = "applied-checksum"
+	appliedOutputKey     = "applied-output"
+	probeStatusesKey     = "probe-statuses"
+	probePeriodKey       = "probe-period"
+	planKey              = "plan"
+	enqueueAfterDuration = "5s"
+)
+
+func Watch(ctx context.Context, applyinator applyinator.Applyinator, connInfo config.ConnectionInfo) error {
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:32]: Does this need to return an error since it can't?

File: pkg/k8splan/watcher.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+			if rawProbeStatusByteData, ok := secret.Data[probeStatusesKey]; ok {
+				if err := json.Unmarshal(rawProbeStatusByteData, &probeStatuses); err != nil {
+					logrus.Errorf("[K8s] error while parsing probe statuses: %v", err)
+					probeStatuses = make(map[string]prober.ProbeStatus, 0)
+				}
+			}
+			// calculate the checksum of the plan from the provided data
+			cp, err := applyinator.CalculatePlan(planData)
+			if err != nil {
+				return secret, err
```
Comment (briandowns) [inline pkg/k8splan/watcher.go:110]: Should this return nil for the secret since there's an error?

Comment (Oats87) [reply pkg/k8splan/watcher.go:110]: the convention is to return the object passed into the `OnChange`, so I don't believe we should be returning `nil`

File: pkg/k8splan/watcher.go (hunk 6 above)
Comment (Oats87) [reply pkg/k8splan/watcher.go]: error removed

**Review by briandowns — APPROVED** [review]

**Pull request merged**